### ISSUE TITLE
Overgenerates and doesn't respect paths=source_relative

### ISSUE THREAD

**Issue by derekperkins** [bufbuild/protoschema-plugins#7]
Testing this next to `protoc-gen-bq-schema`, there were two initial problems that kept me from moving forward:
1. I expected schemas to be opt-in, and only generated when an option was set
`option (gen_bq_schema.bigquery_opts).table_name = "bar_table";`
2. Everything was generated to the same directory, not to the expected destination directories where all my other plugins generate code

`buf.yaml`

```yaml
  - name: bq-schema
    out: gen/go
    opt: paths=source_relative

  - name: bigquery # v0.1.2
    out: gen/go
    opt: paths=source_relative
```

**Comment by Alfus** [comment]
Thanks for the feedback! Agree that both need to be fixed.

**Comment by mfridman** [comment]
> 1. I expected schemas to be opt-in, and only generated when an option was set
>    `option (gen_bq_schema.bigquery_opts).table_name = "bar_table";`

I suppose this should be the default behavior. I did notice there was an upstream issue that requested _"Add option to generate schemas for all messages"_, so ideally we make this configurable?

> Everything was generated to the same directory, not to the expected destination directories where all my other plugins generate code

That's fair. Although we have to be a bit careful about how we implement this.

In your example above `bq-schema` doesn't use `paths=source_relative` (that's usually for Go-generated code with protogen), but instead it does a crude package replacement for `.` with `/`.

My guess, it just so happens that your files are in a directory that matches their package name ([DIRECTORY_SAME_PACKAGE](https://buf.build/docs/lint/rules#directory_same_package)), and if you're using `buf` it'll raise this as a lint error.

TL;DR - there are 2 options

1. Do what `bq-schema` is doing, solely based on package name
2. The equivalent of `paths=source_relative`, where the output filename is derived from the input filename

**Comment by derekperkins** [comment]
> I suppose this should be the default behavior. I did notice there was an upstream issue that requested "Add option to generate schemas for all messages", so ideally we make this configurable?

While configuration is fine, there are so few cases where that feels reasonable. Maybe if you have a dedicated schema repo and there are literally no nested messages? I personally wouldn't think that's worth the effort and added complexity, but maybe that's more common than I I think.

> My guess, it just so happens that your files are in a directory that matches their package name ([DIRECTORY_SAME_PACKAGE](https://buf.build/docs/lint/rules#directory_same_package)), and if you're using buf it'll raise this as a lint error.

The crude replacement seems to work for us. We're using buf and have pretty aggressive linting in plac, but didn't get an error. We're using managed mode if that matters. As is, it's working exactly how we want it to.

```proto
syntax = "proto3";
package nozzle.searches.v1;
```

```yaml
lint:
  use:
    - DEFAULT
    - COMMENT_ENUM
    - COMMENT_MESSAGE
    - COMMENT_RPC
    - COMMENT_SERVICE
    - UNARY_RPC
    - PACKAGE_NO_IMPORT_CYCLE
```

> The equivalent of paths=source_relative, where the output filename is derived from the input filename

The directory structure matches how everything else we run through buf is exported, except the output filename is based on the table name, not the input filename, which is exactly what I expected and what I wanted.

**Comment by derekperkins** [comment]
Speaking of linting errors, right now I just copied `bq_field.proto` and `bq_table.proto` into my repo to reference locally. Is there already or are there plans to have a buf dependency for them?

**Comment by mfridman** [comment]
Thanks for the feedback, I don't think it'd be too hard to plumb a plugin option like `generate_all_messages=true` (or whatever name we come up with). Either way, we'll prioritize only generating when the option is set.

I think the crude replacement is the least surprising option and makes it a drop-in replacement for the upstream plugin. You're using `DEFAULT` which would include the line rule (`DIRECTORY_SAME_PACKAGE`) I mentioned above, which is why the implementation wouldn't matter to you :)

Someone did publish the schema, but I'll bring it up with the team to see if this is something we'd want to add separately.

https://buf.build/liberetech/protoc-gen-bq-schema

**Comment by mfridman** [comment]
Closing this issue as we decided to drop support for the bigquery plugin. See https://github.com/bufbuild/protoschema-plugins/issues/6#issuecomment-2040358836 for more details.
